feat: durable state database support

Roiocam opened the pull request:
## Motivation

resolves: #56 

## tracker

- [x] ~MySQL support~: MySQL has very different ways to generate auto increment column
- [x] Oracle support
- [x] SqlServer support
- [x] update documentation

laglangyue commented:
how to test it,is there any integration testing?

laglangyue commented on core/src/main/scala/org/apache/pekko/persistence/jdbc/state/DurableStateQueries.scala:
```diff
@@ -46,6 +46,7 @@ import pekko.persistence.jdbc.config.DurableStateTableConfiguration
     case "H2"       => new H2SequenceNextValUpdater(profile, durableStateTableCfg)
     case "Postgres" => new PostgresSequenceNextValUpdater(profile, durableStateTableCfg)
+    case "MySQL"    => new MySQLSequenceNextValUpdater(profile, durableStateTableCfg)
     case _          => ???
```
`case _ => ???`
Can we provide a default implementation，it is another feature

pjfanning replied:
unfortunately - this support is not standardised across all db servers - we will need specific versions for different db types

pjfanning commented on core/src/it/scala/org/apache/pekko/persistence/jdbc/integration/ScalaJdbcDurableStateChangesByTagTest.scala:
```diff
@@ -0,0 +1,40 @@
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * This file is part of the Apache Pekko project, which was derived from Akka.
```
* new files should be given the standard ASF header
* This `derived from Akka` header is just for files where an important part of the code is copied from an existing Akka file or a Pekko file that has the `derived from Akka` header.

pjfanning commented:
Would it be possible to add string constants for all these Strings?

```
      case PostgresProfile  => "Postgres"
      case MySQLProfile     => "MySQL"
      case OracleProfile    => "Oracle"
      case SQLServerProfile => "SqlServer"
      case H2Profile        => "H2"
```

And then to remove the direct usage of the strings - use the constants instead.

mdedetrich commented:
> Would it be possible to add string constants for all these Strings?
> 
> ```
>       case PostgresProfile  => "Postgres"
>       case MySQLProfile     => "MySQL"
>       case OracleProfile    => "Oracle"
>       case SQLServerProfile => "SqlServer"
>       case H2Profile        => "H2"
> ```
> 
> And then to remove the direct usage of the strings - use the constants instead.

Definite +1 from me

laglangyue commented on core/src/it/scala/org/apache/pekko/persistence/jdbc/integration/ScalaJdbcDurableStateChangesByTagTest.scala:
```diff
@@ -0,0 +1,50 @@
+}
+
+class OracleScalaJdbcDurableStateStoreQueryTest
+    extends JdbcDurableStateSpec(ConfigFactory.load("mysql-shared-db-application.conf"), Oracle) {
```
should be `oracle-shared-db-application.conf`

Roiocam commented:
It may need another separate PR for MySQL support.

anchann commented:
There's a fix for the MySQL support over in the akka repo:

https://github.com/akka/akka-persistence-jdbc/pull/803

What's the procedure for mirroring these kinds of fixes in the pekko repo? Are there some licensing issues? Is it just a matter of copying the change set and submitting a PR? Is there some automated process?

Roiocam commented:
> There's a fix for the MySQL support over in the akka repo:
> 
> [akka/akka-persistence-jdbc#803](https://github.com/akka/akka-persistence-jdbc/pull/803)
> 
> What's the procedure for mirroring these kinds of fixes in the pekko repo? Are there some licensing issues? Is it just a matter of copying the change set and submitting a PR? Is there some automated process?

Yes, there are license issues, but it is not the most important thing for this PR.

I hadn't seen any code from akka PR before, and then take a short look after you mentioned it. If you review it carefully, you will notice that I already implemented MySQL support on my early commits, but I dropped those after some investigation.

Because currently durable state `global_offset` implementation is based on sequence and Pekko Persistence JDBC will increase sequence after an update or insert. In MySQL, seems likes you can't increase sequence manually, that's why you will see the comment from the Akka maintainer that MySQLSequenceNextValueExpr doesn't work.

anchann commented:
Yeah, I see the comment and the revert now, thank you for pointing out. Sorry, should have been more careful before posting.

laglangyue commented on core/src/main/resources/schema/sqlserver/sqlserver-create-schema.sql:
```diff
@@ -40,3 +40,24 @@ CREATE TABLE "snapshot" (
+        CONSTRAINT [df_global_offset] DEFAULT
+        (NEXT VALUE FOR global_offset),
+    "persistence_id"        VARCHAR(255)   NOT NULL,
+    "revision"              NUMERIC(10, 0) NOT NULL,
```
why not use bigint, because pg use bigint

Roiocam replied:
Just to maintain the same column type as Event Journal.

laglangyue commented on core/src/main/resources/schema/sqlserver/sqlserver-create-schema.sql:
```diff
@@ -40,3 +40,24 @@ CREATE TABLE "snapshot" (
+    "revision"              NUMERIC(10, 0) NOT NULL,
+    "state_payload"         VARBINARY(MAX) NOT NULL,
+    "state_serial_id"       INTEGER        NOT NULL,
+    "state_serial_manifest" VARCHAR(MAX),
```
can we use text, varchar(max) is strange

laglangyue replied:
and pg use VARCHAR(255),it should be this?

Roiocam replied:
Same as `event_journal `, the choice of a new column needs another separate PR and more investigation.

laglangyue commented:
LGTM.
I am not very familiar with persistence. I am just comparing it with the implementation of PG

He-Pin commented on core/src/main/scala/org/apache/pekko/persistence/jdbc/state/DurableStateQueries.scala:
```diff
@@ -31,22 +26,17 @@ import pekko.persistence.jdbc.config.DurableStateTableConfiguration
+    case OracleProfile    => new OracleSequenceNextValUpdater(profile, durableStateTableCfg)
+    // TODO https://github.com/apache/pekko-persistence-jdbc/pull/158
+    // case MySQLProfile     => new MySQLSequenceNextValUpdater(profile, durableStateTableCfg)
+    case _ => throw new UnsupportedOperationException(s"Currently JdbcProfile: <$profile> wasn't support.")
```
Unsupported JdbcProfile <$profile> for durableState.

He-Pin commented on integration-test/src/test/scala/org/apache/pekko/persistence/jdbc/integration/ScalaJdbcDurableStateStoreQueryTest.scala:
```diff
@@ -0,0 +1,54 @@
+class PostgresScalaJdbcDurableStateStoreQueryTest
+    extends JdbcDurableStateSpec(ConfigFactory.load("postgres-shared-db-application.conf"), Postgres) {
+  implicit lazy val system: ActorSystem =
+    ActorSystem("JdbcDurableStateSpec", config.withFallback(customSerializers))
```
as These lines are identical as the old, so we may need keep the lightbend header.

Roiocam replied:
Fair to me.

He-Pin commented on integration-test/src/test/scala/org/apache/pekko/persistence/jdbc/integration/ScalaJdbcDurableStateStoreQueryTest.scala:
```diff
@@ -0,0 +1,54 @@
+  override private[jdbc] def dropAndCreate(schemaType: SchemaType): Unit = {
+    super.dropAndCreate(schemaType)
+    withStatement(stmt => stmt.executeUpdate("""ALTER SEQUENCE global_offset RESTART WITH 1""".stripMargin))
+  }
```
You can swith these tests to a dedicated file with apache header.

Roiocam replied:
That will create a new file, would this current change have a license issue?

He-Pin commented on core/src/main/resources/schema/sqlserver/sqlserver-drop-schema.sql:
```diff
@@ -1,3 +1,4 @@
 DROP TABLE IF EXISTS event_tag;
 DROP TABLE IF EXISTS event_journal;
 DROP TABLE IF EXISTS snapshot;
+DROP TABLE IF EXISTS durable_state;
```
global_offset?

Roiocam replied:
That should be fine, the event_journal's SEQUENCE  doesn't have drop SQL either.

He-Pin commented on core/src/main/scala/org/apache/pekko/persistence/jdbc/state/SequenceNextValUpdater.scala:
```diff
@@ -60,3 +60,31 @@ import slick.sql.SqlStreamingAction
+
+  import profile.api._
+
+  final val nextValFetcher = s"""SELECT NEXT VALUE FOR ${durableStateTableCfg.columnNames.globalOffset}""".stripMargin
```
sql with string template..... :(

Roiocam replied:
Should it be no problem? SQL will be pre-compiled below.

pjfanning replied:
this doesn't need stripMargin call

He-Pin replied:
I think that maybe not safe with string template for sql

He-Pin commented on core/src/main/scala/org/apache/pekko/persistence/jdbc/state/SequenceNextValUpdater.scala:
```diff
@@ -60,3 +60,31 @@ import slick.sql.SqlStreamingAction
+ * INTERNAL API
+ */
+@InternalApi private[jdbc] class OracleSequenceNextValUpdater(profile: JdbcProfile,
+    val durableStateTableCfg: DurableStateTableConfiguration)
```
`val` is not needed.

Roiocam replied:
Let us keep this `val`, otherwise it will complain unused warning, maybe we will use this cfg for the SQL template in the future to avoid binary incompatibility.

pjfanning commented on core/src/main/resources/schema/oracle/oracle-create-schema.sql:
```diff
@@ -55,3 +55,34 @@ BEGIN
+    PERSISTENCE_ID        VARCHAR(255) NOT NULL,
+    REVISION              NUMERIC      NOT NULL,
+    STATE_PAYLOAD         BLOB         NOT NULL,
+    STATE_SERIAL_ID       NUMBER(10) NOT NULL,
```
minor format issue - can you add a space before 'NOT'?

Roiocam replied:
fixed

pjfanning commented:
Unfortunately, the Oracle integration tests are failing due to table name issues.

Roiocam commented:
> Unfortunately, the Oracle integration tests are failing due to table name issues.

Seems like latest commit break this, I will figure out later

pjfanning commented:
@Roiocam can you create an issue for following up with MySQL support - so at least, the issue is there for other users to examine and maybe provide a solution?

Roiocam commented:
> @Roiocam can you create an issue for following up with MySQL support - so at least, the issue is there for other users to examine and maybe provide a solution?

created in #174

pjfanning commented on integration-test/src/test/scala/org/apache/pekko/persistence/jdbc/integration/ScalaJdbcDurableStateStoreQueryTest.scala:
```diff
@@ -0,0 +1,48 @@
+
+  override private[jdbc] def dropAndCreate(schemaType: SchemaType): Unit = {
+    super.dropAndCreate(schemaType)
+    withStatement(stmt => stmt.executeUpdate("""ALTER SEQUENCE global_offset RESTART WITH 1""".stripMargin))
```
the stripMargin call is not needed

He-Pin commented on core/src/main/scala/org/apache/pekko/persistence/jdbc/state/SequenceNextValUpdater.scala:
```diff
@@ -60,3 +60,31 @@ import slick.sql.SqlStreamingAction
+/**
+ * INTERNAL API
+ */
+@InternalApi private[jdbc] class SqlServerSequenceNextValUpdater(profile: JdbcProfile,
```
final class

pjfanning commented on core/src/test/scala/org/apache/pekko/persistence/jdbc/state/scaladsl/JdbcDurableStateSpec.scala:
```diff
@@ -83,7 +83,14 @@ abstract class JdbcDurableStateSpec(config: Config, schemaType: SchemaType) exte
           case Postgres =>
             e shouldBe an[org.postgresql.util.PSQLException]
-          case _ => ???
+          // TODO https://github.com/apache/pekko-persistence-jdbc/pull/158
```
The issue exists and there is a code comment with a link. #174

Roiocam commented:
All suggestions have been apply, there is changes: https://github.com/apache/pekko-persistence-jdbc/pull/158/files/1d3b1e5bf22ac6f7a3cf8c68fd0ad883f3d7bd84..c7bc649641a01c9471754852398d4f44299d9023

pjfanning reviewed (approved):
lgtm
